### PR TITLE
Allow NPCs to repair vehicles

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Features "Allow NPCs to repair vehicles"```

#### Purpose of change
NPCs can now be prompted to repair vehicles in a vehicle repair zone

#### Describe the solution
Same as previous NPC activity PRs I've done - but with some tweaking to allow them to evaluate tools with charges like a welder etc.

#### Describe alternatives you've considered
N/A

#### Additional context
N/A

### CODE REVIEW

File: src/activity_item_handling.cpp [hunk 1]
```diff
@@ -875,6 +875,44 @@ static int move_cost( const item &it, const tripoint &src, const tripoint &dest
     return move_cost_inv( it, src, dest );
 }
 
+static void vehicle_repair_activity( player &p, const tripoint src_loc, int vpindex )
+{
+    vehicle *veh = veh_pointer_or_null( g->m.veh_at( src_loc ) );
+    if( !veh ) {
+        return;
+    } else {
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]:
```suggestion
    }
```
And now you can remove one level of indentation from the remaining code in this function.

File: src/activity_item_handling.cpp [hunk 2]
```diff
@@ -1200,64 +1244,113 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
-                if( abs( veh->velocity ) > 100 || veh->engine_on ) {
-                    continue;
-                }
-                // dont have skill to remove it
-                std::map<skill_id, int> removal_skills = vpinfo.removal_skills;
-                for( const auto &e : removal_skills ) {
-                    bool hasSkill = p.get_skill_level( e.first ) >= e.second;
-                    if( !hasSkill ) {
+                    for( const npc &guy : g->all_npcs() ) {
+                        if( guy.disp_name() != p.disp_name() && guy.activity_vehicle_part_index != -1 &&
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: Why compare the name here? If you want to know whether `guy` and `p` refer to the same NPC, compare their pointers:
```suggestion
                        if( &guy != &p && guy.activity_vehicle_part_index != -1 &&
```

File: src/activity_item_handling.cpp [hunk 3]
```diff
@@ -1200,64 +1244,113 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
+                    if( abs( veh->velocity ) > 100 || veh->engine_on ) {
                         continue;
                     }
+                    // dont have skill to remove it
+                    std::map<skill_id, int> removal_skills = vpinfo.removal_skills;
+                    for( const auto &e : removal_skills ) {
+                        bool hasSkill = p.get_skill_level( e.first ) >= e.second;
+                        if( !hasSkill ) {
+                            continue;
+                        }
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: This loop does nothing. It determines whether a skill is at a certain level or above, but it completely ignores that condition and continues with the next skill.

File: src/activity_item_handling.cpp [hunk 4]
```diff
@@ -1200,64 +1244,113 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
+                    }
+                    item base( vpinfo.item );
+                    if( base.is_wheel() ) {
+                        // no wheel removal yet
+                        continue;
+                    }
+                    const quality_id qual = LIFT;
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: This variable is never used.

File: src/activity_item_handling.cpp [hunk 5]
```diff
@@ -1200,64 +1244,113 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
+                                          TOOL_LIFT_FACTOR );
+                    const bool use_aid = max_lift >= lvl;
+                    const bool use_str = p.can_lift( base );
+                    if( !( use_aid || use_str ) ) {
+                        continue;
+                    }
+                    const auto &reqs = vpinfo.removal_requirements();
+                    const std::string ran_str = random_string( 10 );
+                    const requirement_id req_id( ran_str );
+                    requirement_data::save_requirement( reqs, req_id );
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: What does that even attempt to do? The requirement is stored in the list of all requirements using a random id.

But that id is never stored anywhere else. So there is no way of getting this requirement back (the ID is random!).

File: src/activity_item_handling.cpp [hunk 6]
```diff
@@ -1200,64 +1244,113 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
+                            continue;
+                        }
+                    }
+                    const auto &reqs = vpinfo.repair_requirements();
+                    const std::string ran_str = random_string( 10 );
+                    const requirement_id req_id( ran_str );
+                    requirement_data::save_requirement( reqs, req_id );
+                    std::vector<tripoint> points_to_check;
+                    for( const auto elem : g->m.points_in_radius( src_loc, PICKUP_RANGE - 1 ) ) {
+                        points_to_check.push_back( elem );
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: This is even more nonsensical. The variable `points_to_check` is set, but never read.

File: src/activity_item_handling.cpp [hunk 7]
```diff
@@ -2407,15 +2513,22 @@ void generic_multi_activity_handler( player_activity &act, player &p )
...
                 // we already checked this in can_do_activity() but check again just incase.
                 if( !veh ) {
                     p.activity_vehicle_part_index = 1;
                     continue;
                 }
                 const vpart_info &vpinfo = veh->part_info( p.activity_vehicle_part_index );
-                const auto &reqs = vpinfo.removal_requirements();
+                requirement_data reqs;
+                if( reason == NEEDS_VEH_DECONST ) {
+                    const requirement_data &local_req = vpinfo.removal_requirements();
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]:
```suggestion
                    reqs = vpinfo.removal_requirements();
```
Same below. No need for a new variable.

File: src/activity_item_handling.cpp [hunk 8]
```diff
@@ -1200,64 +1244,113 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
+                    for( const auto &e : repair_skills ) {
+                        bool hasSkill = p.get_skill_level( e.first ) >= e.second;
+                        if( !hasSkill ) {
+                            continue;
+                        }
+                    }
+                    const auto &reqs = vpinfo.repair_requirements();
+                    const std::string ran_str = random_string( 10 );
+                    const requirement_id req_id( ran_str );
+                    requirement_data::save_requirement( reqs, req_id );
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: Same here with `req_id`, The requirement is stored, but the **random** id is not stored anywhere.

Btw. even if the ID was stored somewhere else, the requirement is not stored in the save, so after reloading the game, the ID will not be valid anymore.

File: src/activity_item_handling.cpp [hunk 9]
```diff
@@ -1200,64 +1244,113 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
-                    continue;
-                }
-                for( const npc &guy : g->all_npcs() ) {
-                    if( guy.disp_name() != p.disp_name() && guy.activity_vehicle_part_index != -1 &&
-                        guy.activity_vehicle_part_index == vpindex ) {
+            if( act == activity_id( "ACT_VEHICLE_DECONSTRUCTION" ) ) {
+                // find out if there is a vehicle part here we can remove.
+                std::vector<vehicle_part *> parts = veh->get_parts_at( src_loc, "", part_status_flag::any );
+                for( vehicle_part *part_elem : parts ) {
+                    const vpart_info &vpinfo = part_elem->info();
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: Unused.

File: src/activity_item_handling.cpp (hunk 2 above)
Comment (davidpwbrown) [reply src/activity_item_handling.cpp]: guy is an npc and p is a player, in this context. Worth a cast to player for guy, to compare?

File: src/activity_item_handling.cpp (hunk 9 above)
Comment (davidpwbrown) [reply src/activity_item_handling.cpp]: vpinfo? that is used,  in `std::map<skill_id, int> removal_skills = vpinfo.removal_skills;` a few lines down.

**Comment by davidpwbrown** [comment]
Ok I'm gonna rework the storing part indexes, cos they are too easily invalidated by other NPCs or monsters potentially smashing the part there etc.

File: src/activity_item_handling.cpp [hunk 10]
```diff
@@ -1187,77 +1229,119 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
+                    if( std::find( already_working_indexes.begin(), already_working_indexes.end(),
+                                   vpindex ) != already_working_indexes.end() ) {
+                        continue;
+                    }
+                    // dont have skill to remove it
+                    std::map<skill_id, int> repair_skills = vpinfo.repair_skills;
+                    for( const auto &e : repair_skills ) {
+                        bool hasSkill = p.get_skill_level( e.first ) >= e.second;
+                        if( !hasSkill ) {
+                            continue;
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: This skill check loop is still non-functional (like the other one). Hint: If the code wasn't duplicated all over the place, but in a simple function that had been called from here, you would not need make sure to update it all over the place.

File: src/activity_item_handling.cpp [hunk 11]
```diff
@@ -1187,77 +1229,119 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
+                    const int max_lift = p.best_nearby_lifting_assist( src_loc );
+                    const int lvl = ceil( units::quantity<double, units::mass::unit_type>( base.weight() ) /
+                                          TOOL_LIFT_FACTOR );
+                    const bool use_aid = max_lift >= lvl;
+                    const bool use_str = p.can_lift( base );
+                    if( !( use_aid || use_str ) ) {
+                        continue;
+                    }
+                    const auto &reqs = vpinfo.removal_requirements();
+                    const std::string ran_str = random_string( 10 );
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: This seems to be unused now.

File: src/activity_item_handling.cpp [hunk 12]
```diff
@@ -1187,77 +1229,119 @@ static activity_reason_info can_do_activity_there( const activity_id &act, playe
...
-    if( act == activity_id( "ACT_VEHICLE_DECONSTRUCTION" ) ) {
-        if( g->m.getlocal( g->u.activity.placement ) == src_loc ) {
-            return activity_reason_info::fail( ALREADY_WORKING );
-        }
+    if( act == activity_id( "ACT_VEHICLE_DECONSTRUCTION" ) ||
+        act == activity_id( "ACT_VEHICLE_REPAIR" ) ) {
+        std::vector<int> already_working_indexes;
         for( const npc &guy : g->all_npcs() ) {
-            if( g->m.getlocal( guy.activity.placement ) == src_loc || guy.pos() == src_loc ) {
+            if( guy.disp_name() == p.disp_name() ) {
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: What actually happens when the player has two followers with the same name? Should compare the pointers directly. (Comparing pointer to npc and pointer to player will work fine.)

Comment (davidpwbrown) [reply src/activity_item_handling.cpp]: Ah hah, didnt know if comparing npc to player would work.  That makes more sense. thanks.

**Comment by davidpwbrown** [comment]
Fixed error in time taken for vehicle activity, improved part lock that other NPCs are working on, by  checking if its the same vehicle first. 

Will try and serialize the custom requirements in a seperate PR. this seems work as intended now, pending any further required code changes.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/new-idea-for-cdda/21458/5

**Pull request merged**